### PR TITLE
Add support for Cyclone DDS

### PR DESCRIPTION
**Pull request by ruffsl**
Context: ros2/ros2#786
Related: https://github.com/ros2/ci/pull/340

### CODE REVIEW

File: ros2.repos [hunk 1]
```diff
@@ -147,6 +151,10 @@ repositories:
     type: git
     url: https://github.com/ros2/console_bridge_vendor.git
     version: master
+  ros2/rmw_cyclonedds:
```
Comment (dirk-thomas) [inline ros2.repos]: Please insert in alphabetical order.

Comment (ruffsl) [reply ros2.repos]: @dirk-thomas Fixed.

**Comment by dirk-thomas** [comment]
CI builds with only CycloneDDS - just to check that the Jenkins job parameter works as intended and as a base line:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8226/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4198/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6706/) (needs CUnit to be installed)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8111/) (needs CUnit to be installed)

CI build without CycloneDDS - just to check that the Jenkins job parameter works as intended:
* macOS: [Build Status](https://ci.ros2.org/job/ci_osx/6707/)

**Comment by dirk-thomas** [comment]
Waiting for the macOS and Windows Jenkins nodes to have the necessary prerequisites.

**Comment by cottsay** [comment]
I installed CUnit and Maven on the OSX workers, but the CycloneDDS build failed: https://ci.ros2.org/job/ci_osx/6719/console#console-section-22

I'm still looking for a good way to get CUnit on the Windows workers...

**Comment by dirk-thomas** [comment]
> I installed CUnit and Maven on the OSX workers, but the CycloneDDS build failed: https://ci.ros2.org/job/ci_osx/6719/console#console-section-22

@eboasson Can you please take a look at this linker error:

```
Undefined symbols for architecture x86_64:
  "_fmemopen", referenced from:
      _CU_TestProxy_dds_log_only_log_file in log.c.o
      _CU_TestProxy_dds_log_same_file in log.c.o
      _CU_TestProxy_dds_log_no_sink in log.c.o
ld: symbol(s) not found for architecture x86_64
```

**Comment by eboasson** [comment]
@dirk-thomas and @cottsay I expect the macOS build will be fine now. The ``fmemopen`` dependency is in a single test case, so pretty easy to avoid.

Getting CUnit on Windows is something we do with [Conan](https://conan.io). I don't know if that would work for you, but I would think you would be able to install Conan without any trouble given that you use Python a lot and you could then perhaps just use the same as we do on Travis:

https://github.com/eclipse-cyclonedds/cyclonedds/blob/50b0611ad1d6a1589cf5d8e2abb424ed83cbc7c6/.travis.yml#L161-L162

and

https://github.com/eclipse-cyclonedds/cyclonedds/blob/50b0611ad1d6a1589cf5d8e2abb424ed83cbc7c6/.travis.yml#L181-L184

**Comment by dirk-thomas** [comment]
> I expect the macOS build will be fine now.

I just ran a build with https://github.com/eclipse-cyclonedds/cyclonedds/commit/50b0611ad1d6a1589cf5d8e2abb424ed83cbc7c6 and the result is still the same: [Build Status](https://ci.ros2.org/job/ci_osx/6724/)

**Comment by eboasson** [comment]
@dirk-thomas it looks like the problem now is that my test only correctly detects whether ``fmemopen`` is available as long as the Xcode version is new enough to define the version macro for macOS 10.13. Of course it wasn't defined at the time of macOS 10.12, and of course I forgot to account for that ...

I now have a PR (https://github.com/eclipse-cyclonedds/cyclonedds/pull/265) on Cyclone that fixes the test and adds a macOS 10.12 + Xcode 9 build to the set of automated builds. It's waiting for a review, so the chances of the update being picked up in today's build are minimal (perhaps already gone completely, depending on when the build gets started).

**Comment by eboasson** [comment]
@dirk-thomas Cyclone master now builds & tests on macOS 10.12 in our Travis setup, so it is definitely worth another go.

**Comment by dirk-thomas** [comment]
The latest build looks much better: [Build Status](https://ci.ros2.org/job/ci_osx/6737/)

**Comment by dirk-thomas** [comment]
> I'm still looking for a good way to get CUnit on the Windows workers...

@cottsay Any update on updating the Windows workers?

**Comment by cottsay** [comment]
> @cottsay Any update on updating the Windows workers?

Current plan is to create a chocolatey package for CUnit. We're not planning to install another package manager for this. I also need to figure out what it will take for Maven.

**Comment by cottsay** [comment]
I manually created a chocolatey package for CUnit, but I'm still tweaking things to prepare it for wider distribution and inclusion in the installation instructions. I also installed Maven and switched our Windows workers from the JRE to the full JDK, which was also required for Cyclone.

In any case, I got the first build out, which failed during compilation: [Build Status](https://ci.ros2.org/view/All/job/test_ci_windows/72/)

Please take a look, @eboasson.

**Comment by eboasson** [comment]
@cottsay the joys of using Microsoft's toolchain ...

I figure that the Conan package we use on Travis/AppVeyor is a good source of information for using CUnit on "modern" Windows. There, the following patch is applied to the CUnit sources:

```patch
--- source_subfolder/CUnit/Headers/CUnit.h.in	2014-04-24 02:54:45.000000000 +0200
+++ source_subfolder/CUnit/Headers/CUnit.h.in.patched	2018-06-22 18:21:11.755600964 +0200
@@ -102,7 +102,9 @@
 #    define CU_EXPORT
 #  endif
 #  ifdef _MSC_VER
-#    define snprintf _snprintf
+#    if _MSC_VER < 1900
+#      define snprintf _snprintf
+#    endif
 #  endif
 #else
 #  define CU_EXPORT
```

(The package info is at https://bintray.com/bincrafters/public-conan/cunit%3Abincrafters#)

It does look to me like that patch will indeed solve the errors messages you got:
```
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\ucrt\stdio.h(1933): warning C4005: 'snprintf': macro redefinition (compiling source file C:\J\workspace\test_ci_windows\ws\src\eclipse-cyclonedds\cyclonedds\src\core\ddsc\tests\instance_get_key.c) [C:\J\workspace\test_ci_windows\ws\build\cyclonedds\src\core\ddsc\tests\cunit_ddsc.vcxproj]
  C:\ProgramData\chocolatey\lib\cunit\include\CUnit\CUnit.h(105): note: see previous definition of 'snprintf' (compiling source file C:\J\workspace\test_ci_windows\ws\src\eclipse-cyclonedds\cyclonedds\src\core\ddsc\tests\instance_get_key.c)
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\ucrt\stdio.h(1935): fatal error C1189: #error:  Macro definition of snprintf conflicts with Standard Library function declaration (compiling source file C:\J\workspace\test_ci_windows\ws\src\eclipse-cyclonedds\cyclonedds\src\core\ddsc\tests\instance_get_key.c) [C:\J\workspace\test_ci_windows\ws\build\cyclonedds\src\core\ddsc\tests\cunit_ddsc.vcxproj]
```

**Comment by cottsay** [comment]
Okay, progress. Here are the current blockers I'm aware of:
- [ ] eclipse-cyclonedds/cyclonedds#291
- [ ] ros2/choco-packages#15
- [ ] Release the CUnit Choco package for Windows
- [ ] eclipse-cyclonedds/cyclonedds#292
- [ ] ros2/ros2_documentation#369

**Comment by cottsay** [comment]
AFAIK, this should be good to go.

Fresh CI after merging with `master`:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8552/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4424/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6972/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8465/)

**Comment by mjcarroll** [comment]
@cottsay want to take another stab at this one?  Windows hard fails at compilation.

**Comment by cottsay** [comment]
@eboasson, this looks like an RMW bug to me:
```
C:\J\workspace\ci_windows\ws\src\ros2\rmw_cyclonedds\rmw_cyclonedds_cpp\src\rmw_node.cpp(505,5): fatal error C1012: unmatched parenthesis: missing ')' 
```

**Comment by eboasson** [comment]
@cottsay indeed, the problem is that the MSVC preprocessor doesn't accept the use of the ternary operator in a preprocessor conditional. I guess GCC and Clang supporting it must be an extension.

PR https://github.com/ros2/rmw_cyclonedds/pull/58 fixes the issue (on Eloquent on macOS 10.15, Dashing on Ubuntu and a minimalistic Eloquent build on Win10/VS2019).

**Comment by cottsay** [comment]
C'mon, c'mon...
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8568/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4443/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6989/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8486/)

**Comment by cottsay** [comment]
Alright, we have a handful of test failures on the POSIX systems, but the Windows tests are in rough shape. I think that something fundamental is malfunctioning that is causing pretty well all of the tests to fail. The tests are retried 10 times, so the build is taking a really long time to fail and will likely just time out.

Since CycloneDDS is only turned on for some of the builds, the handful of test failures might be OK to merge and then iterate on, but I think we need a plan for the Windows failures before we merge.

**Comment by dirk-thomas** [comment]
We have no nightly job building / testing CycloneDDS on Windows. Therefore I think this should be merged now.

**Comment by eboasson** [comment]
The test failures on POSIX match with some gaps in QoS support in Cyclone and are to be expected for now (there's progress on implementing them, e.g., https://github.com/eclipse-cyclonedds/cyclonedds/pull/274), but the Windows troubles are a complete surprise.

My attempts at getting my hands on a reliably functioning Windows 10 (virtual) box that can build enough of ROS2 for me to investigate a bit over the weekend have come to a ignominious end. I've become convinced Windows has gained some level of sentience and now responds to my aversion to its ways and inner belief system by refusing to retain ``cmake`` in its file system — though perhaps that just means it shares my dislike for cmake — and by shutting down on me. Therefore it'll have to wait until next week ...

Should anyone be interested in giving it a shot: is everything failing or is it just the tests that require communication? If it is the latter, it might be good to check what Cyclone is actually doing. [Enabling tracing](https://github.com/eclipse-cyclonedds/cyclonedds#configuration) and then running just one of the tests might reveal something.

**Comment by eboasson** [comment]
Progress: the problem is (almost certainly) that during shutdown it needs to stop all threads but one thread is not woken up from its blocking read. That blocking read is on a socket used for receiving multicasts and the thread driving the shutdown tries to wake it by sending a multicast. The socket is subscribed to it and yet the datagram doesn't arrive, for reasons as yet unknown. Somehow, the kernel simply refuses to deliver the packet. It is not something as trivial as random packet loss because it retries every 1s.

From experimentation here by @dennis-adlink and (to a lesser extent) myself, it never seems to fail if the loopback interface is selected. Other networks, it seems to vary with whether the machine was recently booted and on whether VirtualBox and/or Wireshark are running. Basically, after booting Windows it works fine, but that doesn't mean it will still work a few hours later, despite nothing "interesting" having happened.

The code explicitly sets the outgoing multicast interface to whatever interface it is listening on (and that's been working fine for a decade or so), and yet Wireshark shows the source address as 127.0.0.1 instead — but then I'm not sure whether to trust Wireshark in a case like this:

upd_mc_fail

Come to think of it, it may also have something to do with firewalls. I do know that on macOS with Little Snitch installed, similar things can happen — and just so Linux doesn't feel left out, I know of a system using OpenSplice on Linux where the source IP address in the packet doesn't match the network interface it is sent out on! So in any case, multicast weirdness is not unheard of.

Given the observations, there are several workarounds:
* read data from all sockets in a single thread, because the way it multiplexes the sockets allows it to wake up the thread without sending a packet (``export CYCLONEDDS_URI=<Internal><MultipleReceiveThreads>false</></>`` will do the trick);
* disabling use of multicast for data, because then there is no thread in a blocked read on a socket used for multicast (``export CYCLONEDDS_URI=<General><AllowMulticast>spdp</></>``);
* or, apparently, using the loopback interface (merging https://github.com/ros2/rmw_cyclonedds/pull/60 will make that easy, or else ``export CYCLONEDDS_URI=<General><NetworkInterfaceAddress>localhost</></>``).

If it is easy to set an environment variable by way of an experiment, perhaps it would make sense to try one of the above? Not that it isn't important to understand the root cause ...

**Comment by nuclearsandwich** [comment]
> We have no nightly job building / testing CycloneDDS on Windows. Therefore I think this should be merged now.

rmw_cyclone_dds gets picked up by the extra_rmw_release jobs. I've opened a ros2/ci#352 to exclude it until it passes.

**Comment by dirk-thomas** [comment]
> If it is easy to set an environment variable by way of an experiment, perhaps it would make sense to try one of the above?

Yes, that is trivial to do.

**Pull request merged**